Few daemon duplication refactor

vdemeester opened the pull request:
Few small refactoring to limit duplication where I thought I could. For some part this is more a try to get opinion than a definite refactoring. It is separate in separate commit to ease the review and also to remove if the refactoring doesn't make sense.

The refactor happens mainly in `cluster` package and at some part of the `daemon` one.

/cc @aaronlehmann @tonistiigi @AkihiroSuda @yongtang @cpuguy83 

Hopefully this does not break anything 👼

🐸

vdemeester commented on daemon/cluster/cluster.go:
```diff
@@ -1570,8 +1529,7 @@ func (c *Cluster) CreateNetwork(s apitypes.NetworkCreateRequest) (string, error)
 
-// RemoveNetwork removes a cluster network.
-func (c *Cluster) RemoveNetwork(input string) error {
+func (c *Cluster) updateClusterObject(updateFn func(ctx context.Context, state nodeState) error) error {
```
I think we could do better than that 👼

allencloud replied:
I am not sure if `updateClusterObject` does not contain the following codes
```
        state := c.currentNodeState()
 	if !state.IsActiveManager() {
 		return c.errNoManager(state)		
 	}
```
will the handler return an error of `this node is not part of swarm`.

Then a docker with no swarm mode will response to cluster update action which may break api.

vdemeester replied:
The github diff doesn't really show it properly, but `updateClusterObject` does contain this piece of code 😉

aaronlehmann replied:
I would prefer a name like `lockedManagerAction`.

vdemeester commented on daemon/container.go:
```diff
@@ -54,6 +54,19 @@ func (daemon *Daemon) GetContainer(prefixOrName string) (*container.Container, e
 	return daemon.containers.Get(containerID), nil
 }
 
+func (daemon *Daemon) GetcontainerCondition(prefixOrName string, conditions ...func(*container.Container) error) (*container.Container, error) {
```
I don't like the name I used 👼 I should change it 😄

yongtang replied:
The case may be changed, e.g., `s/GetcontainerCondition/GetContainerCondition/`, ?

aaronlehmann replied:
Agreed, I see what you're going for with `GetcontainerCondition` but I don't care for it. What about `CheckContainer`?

vdemeester replied:
@aaronlehmann for the name of the method, or as a method that would only do the checks ?

aaronlehmann replied:
The name of the method.

vdemeester commented on daemon/container_operations.go:
```diff
@@ -829,16 +829,24 @@ func (daemon *Daemon) disconnectFromNetwork(container *container.Container, n li
+				logrus.Warnf("error detaching from network %s: %v", network.ID(), err)
 			}
 		}
+		attributes := map[string]string{
```
This was not on the in the initial "call" (above) and I feel that's a miss.

vdemeester commented on daemon/graphdriver/counter.go:
```diff
@@ -24,29 +24,19 @@ func NewRefCounter(c Checker) *RefCounter {
+	})
+}
+
+func (c *RefCounter) operation(path string, infoOp func(minfo *minfo)) int {
```
This one I'm not sure about 😛

AkihiroSuda replied:
I wonder it is too much to have general func here.
I think `func (c *RefCounter) incdec(path string, inc bool)) int` is ok

LK4D4 commented:
@vdemeester gofmt issue

vdemeester commented:
@LK4D4 yeah saw that, should be fixed 👼

AkihiroSuda commented on daemon/graphdriver/counter.go:
```diff
@@ -24,29 +24,19 @@ func NewRefCounter(c Checker) *RefCounter {
 // Decrement decreases the ref count for the given id and returns the current count
 func (c *RefCounter) Decrement(path string) int {
+	return c.operation(path, func(minfo *minfo) {
+		minfo.count--
```
Probably need zero check here
https://github.com/docker/docker/pull/30811

aaronlehmann commented:
There are still `gofmt` issues.

aaronlehmann commented on daemon/daemon.go:
```diff
@@ -879,33 +879,30 @@ func (daemon *Daemon) Unmount(container *container.Container) error {
 
 // V4Subnets returns the IPv4 subnets of networks that are managed by Docker.
 func (daemon *Daemon) V4Subnets() []net.IPNet {
```
What about refactoring `V4Subnets` and `V6Subnets` into a single function that returns both slices? They are both called in the same place: https://github.com/docker/docker/blob/584e399d9632200e6c8a636339c8aa6347a57fca/daemon/cluster/listen_addr.go#L164-L166

vdemeester replied:
ah you're right 👼

aaronlehmann commented on daemon/daemon_unix.go:
```diff
@@ -1265,19 +1265,21 @@ func (daemon *Daemon) initCgroupsPath(path string) error {
+	return nil
+}
+
+func createCPURealTimeFileIfRequired(sysinfoPresent bool, configValue int64, file string, path string) error {
```
`maybeCreateCPURealTimeFile`?

aaronlehmann commented on daemon/list.go:
```diff
@@ -382,6 +346,27 @@ func (daemon *Daemon) foldFilter(config *types.ContainerListOptions) (*listConte
+func portOp(key string, filter map[nat.Port]bool) func(value string) error {
+	return func(value string) error {
+		if strings.Contains(value, ":") {
+			return fmt.Errorf("filter for '%s' should not contain ':': %v", key, value)
```
Why `%s` for `key` and `%v` for `value` when they are both strings?

vdemeester replied:
Didn't change the code inside (just extracted it), but you are right, we could use `%s` for both, I'll update :wink:

aaronlehmann commented on daemon/list.go:
```diff
@@ -382,6 +346,27 @@ func (daemon *Daemon) foldFilter(config *types.ContainerListOptions) (*listConte
+		proto, port := nat.SplitProtoPort(value)
+		start, end, err := nat.ParsePortRange(port)
+		if err != nil {
+			return fmt.Errorf("error while looking up for %s %v: %s", key, value, err)
```
Why `%s` for `key` and `%v` for `value` when they are both strings?

aaronlehmann commented on daemon/list.go:
```diff
@@ -382,6 +346,27 @@ func (daemon *Daemon) foldFilter(config *types.ContainerListOptions) (*listConte
+		for i := start; i <= end; i++ {
+			p, err := nat.NewPort(proto, strconv.FormatUint(i, 10))
+			if err != nil {
+				return fmt.Errorf("error while looking up for %s %v: %s", key, value, err)
```
Why `%s` for `key` and `%v` for `value` when they are both strings?

vdemeester commented:
Updated 👼

aaronlehmann commented on daemon/daemon.go:
```diff
@@ -879,40 +879,28 @@ func (daemon *Daemon) Unmount(container *container.Container) error {
+		v4infos, v6infos := managedNetwork.Info().IpamInfo()
+		for _, info := range v4infos {
+			if info.IPAMData.Pool != nil {
+				v4Subnets = append(v6Subnets, *info.IPAMData.Pool)
```
Wrong argument to `append`.

vdemeester replied:
ah damn :sweat_smile:

AkihiroSuda commented on daemon/cluster/networks.go:
```diff
@@ -211,51 +208,38 @@ func (c *Cluster) DetachNetwork(target string, containerID string) error {
+		if err != nil {
+			return err
+		}
+		resp = r
```
nit: you don't need `r`

AkihiroSuda commented on daemon/cluster/services.go:
```diff
@@ -59,207 +59,181 @@ func (c *Cluster) GetServices(options apitypes.ServiceListOptions) ([]types.Serv
+		resp := &apitypes.ServiceUpdateResponse{}
+
+		// pin image by digest
+		if os.Getenv("DOCKER_SERVICE_PREFER_OFFLINE_IMAGE") != "1" {
```
similar to L110-121, can we deduplicate? (in future PR)

tonistiigi commented on daemon/container.go:
```diff
@@ -54,6 +54,16 @@ func (daemon *Daemon) GetContainer(prefixOrName string) (*container.Container, e
 }
 
+// CheckContainer make sure the specified container validates the specified conditions
+func (daemon *Daemon) CheckContainer(container *container.Container, conditions ...func(*container.Container) error) error {
```
we could probably leave this method private

vdemeester replied:
right 👼

tonistiigi commented on daemon/container_operations_unix.go:
```diff
@@ -58,32 +58,34 @@ func (daemon *Daemon) setupLinkedContainers(container *container.Container) ([]s
+func containerIsRunning(c *container.Container) error {
 	if !c.IsRunning() {
-		return nil, fmt.Errorf("cannot join PID of a non running container: %s", containerID)
+		return fmt.Errorf("cannot join IPC of a non running container: %s", c.ID)
```
This message would be also shown on pidns error

vdemeester replied:
ah you are right, I need to handle errors differently 👼

vdemeester commented:
@tonistiigi updated 👼

tonistiigi commented:
LGTM

AkihiroSuda reviewed (approved):
LGTM, my above comments are nitpicky :smile:


I think this PR can be merged without squashing, but deferring to somebody else for confirmation :sweat_smile:

thaJeztah commented on daemon/container.go:
```diff
@@ -54,8 +54,8 @@ func (daemon *Daemon) GetContainer(prefixOrName string) (*container.Container, e
-// CheckContainer make sure the specified container validates the specified conditions
-func (daemon *Daemon) CheckContainer(container *container.Container, conditions ...func(*container.Container) error) error {
+// checkContainer make sure the specified container validates the specified conditions
+func (daemon *Daemon) checkContainer(container *container.Container, conditions ...func(*container.Container) error) error {
```
These checkContainer changes are in the wrong commit; should be in f5c490ac633c25ef5f83e99e524a954d50801fb5

thaJeztah commented on daemon/container_operations_unix.go:
```diff
@@ -60,23 +60,23 @@ func (daemon *Daemon) getIpcContainer(container *container.Container) (*containe
+		return nil, errors.Wrapf(err, "cannot join IPC of a non running container: %s", container.ID)
 	}
-	return container, daemon.CheckContainer(container, containerIsRunning, containerIsNotRestarting)
+	return container, daemon.checkContainer(container, containerIsRunning, containerIsNotRestarting)
```
Same for these :smile:

vdemeester committed:
Remove daemon.VXSubnets duplicate code

Refactor daemon.V4Subnets and daemon.V6Subnets to limit duplication

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester committed:
Refactor endpoint*FromGRPC to limit duplication

endpointSpecFromGRPC and endpointFromGRPC do the exact same thing for
endpoint{,Spec}.Ports, let's extract that to a method.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester commented:
@thaJeztah updated 👼

thaJeztah commented on daemon/container_operations_unix.go:
```diff
@@ -69,14 +69,14 @@ func (daemon *Daemon) getPidContainer(container *container.Container) (*containe
 func containerIsRunning(c *container.Container) error {
 	if !c.IsRunning() {
-		return fmt.Errorf("cannot join IPC of a non running container: %s", c.ID)
+		return errors.Errorf("container %s is not running", c.ID)
```
Actually, looks like these changes also belong in 25caa5e9194db2cee946e9ef98557dca9976f6d3

https://github.com/docker/docker/pull/30843/commits/25caa5e9194db2cee946e9ef98557dca9976f6d3#r102823906

vdemeester committed:
Introduce a CheckContainer to remove duplication

Some methods need to get a container *and* validate some conditon on
these (is the container running, …). The CheckContainer allows
to do that and helps remove some duplication.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester committed:
Refactor publish/expose filter to remove duplication

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester committed:
Extract common code from disconnectFromNetwork and releaseNetwork

Both method are trying to detach the container from a cluster
network. The code is exactly the same, this removes the duplication.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester committed:
Reduce duplication in graphdriver

Removes some duplication in counter.go and proxy.go

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester committed:
Refactor cpu-realtime file creation to remove duplication

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester committed:
Add a lockedManagerAction method to Cluster…

… in order to remove duplication.
Each time we update a cluster object, we do some common
operations (lock, verify it's on a manager, get the request context,
and the update). This introduce a method and refactor few
update/remove method that allows to duplicate less code.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

thaJeztah reviewed (approved):
LGTM
